ls1intum/Artemis
# `Notifications`: Fix loading and display issue when opening the notification settings

### Checklist
#### General
#### Server

### Motivation and Context

The issue was that sometimes the notification settings UI did not render the checkboxes for the individual settings correctly, i.e. the checkboxes were missing -> functionality to (de)activate options was broken.
![image](https://user-images.githubusercontent.com/65814168/141649038-b0a094cf-f4c9-4cfc-a654-49414b6f865d.png)

I fixed this issue by making sure the retrieved user data from the DB is checked and corrected if there are issues.

### Description

My problem fixing this issue was/is that I personally struggle to find users where this error occurs (locally or e.g. on TS3).

Thanks to some fellow developers I noticed that this could be easily fixed if you modified the working checkboxes and pressed the save button, afterwards the settings worked again. (But this is a workaround and not a fix)

We inspected the network traffic and found out that the number of settings retrieved from the server was less than the number of default settings (i.e. the number of settings that should be rendered) -> the problem had to be on the server side.

I inspected the code and found out that I do not check the settings after I retrieve them from the DB for correctness. 
There were two cases:
a) If the user has not modified/saved his settings there are no settings entries in the DB -> we return the default settings. (all ok)
b) (here comes the issue) If the user has entries in the DB simply return them. 
What is the problem for b) ? -> At the client side if the user changes his settings and presses "save" all settings (also unchanged) are sent and saved in the DB.
Example of the problem:
Lets say we have 10 settings. The user modifies them and saves -> DB has 10 entries for this user. -> After some time we add a new setting (so now the default has 11 settings) -> But as long as the user does not change/save his settings again the server will only retrieve 10 instead of 11 entries -> error in the html template.

How I fixed this:
I added a server-side check that looks at the retrieved settings and the default settings (based on their unique identifier (notification setting id (String)). If they are not equal delete all notification settings entries (loaded settings) from the user in the DB. Merge the loaded with the default settings and save them in the DB to make sure the DB is up to date (remove outdated entries). If the settings were equal simply return the retrieved settings.

The issue appeared (was triggered) most likely because I e.g. changed the name of a settingsId (from "startedOrCreated" to "released") and we added new METIS (post) settings.

Quick local code example (right side is the DB view. I added a wrong entry called "debugging.test.A" for user 2 which will be removed)
![image](https://user-images.githubusercontent.com/65814168/141649856-b62c3e9c-940a-4ebf-8bcf-10fbf1686af2.png)
![image](https://user-images.githubusercontent.com/65814168/141649860-293b924a-9773-41f8-a35e-cce643f2057e.png)
![image](https://user-images.githubusercontent.com/65814168/141649863-a5642a72-2004-428a-9a3f-ab19f0a532ac.png)


### Steps for Testing
Check if your settings are displayed correctly.
As mentioned before this might be very hard to test because these are quite rare edge-cases (at least when I tried to find examples)

The best thing you can do is to look at the code and make sure it is valid. Thank you in advance!

### Review Progress

#### Code Review
#### Manual Tests

### Test Coverage
NotificationSettingsService > 90%
NotificationSettingsResource > 90% (methods 100%)
